content/wayland-rs-v0.20.markdown

```diff
@@ -0,0 +1,158 @@
+some points not very different from manually re-implementing a v-table. Other issues more
+specific to smithay resulted in a large accumulation of template parameters on types that need
+to be named and stored. All these issues made the API quite tedious to use overall.
+
+As a result, for this new version I decided to embrace trait objects, `Rc`, and `Arc`.
+Additionaly, the API redesign was made focusing on *the spirit* of the wayland protocol,
+rather than trying to stay close to *the C API*.
+
+To list a few clear gains from this rework:
+
+- The scanner (crate that generate API code from the XML protocol files) almost no longer
+  needs to generate `unsafe { }` blocks in its generated code
+- The long standing issue of handling both `Send` and non-`Send` shared state is finally
+  solved (see next section)
+- The ergonomics and flexibility of specifying callbacks for wayland events has been greatly
+  improved
+- Overall better interaction with objects from the C world, if wayland-rs is used in
+  combination with C libraries
+
+## The new callback specification
+
+The most major API change is on the way callbacks are handled. Previously you had to provide
+a set of freestanding functions that would be given access to some state data for each event,
+with the difficulties of sharing state between an undeterminate set of callbacks a priori...
+
+Now each time a new protocol object is created, you'll receive it as a `NewProxy`/`NewResource`
+object (depending on whether you are client side or server side), and will have to implement
+it before accessing the real `Proxy`/`Resource` and being able to use it. Implementing an
+object consists in providing a type implementing the appropriate `Implementation<Meta, Msg>`
+trait.
+
+This trait has a very simple definition:
+
+```rust
+pub trait Implementation<Meta, Msg> {
+    fn receive(&mut self, msg: Msg, meta: Meta);
+}
+```
+
+Here the `Meta` type parameter represent some metadata associated with the message you
+receive (often a `Proxy`/`Resource` handle to the wayland object receiving the message),
+while `Msg` is a type representing the message itself (often an enum of the possible wayland
+messages this object can receive).
+
+The convenience is that `Implementation<Meta, Msg>` is automatically implemented for all
+type that implement `FnMut(Msg, Meta)`. As such, you can easily provide closures as
+implementations, with all the ergonomics of capturing values and code compacity that
+implies. The only catch being that the implementations must be `'static`, so you'll need
+to rely on `Arc` or `Rc` to share data. The runtime cost implied should be negligible however,
+the wayland protocol part should hardly be a bottleneck for any application.
+
+There are also two ways to implement an object, the default one requiring from the 
+implementation to be `Send`, and a secondary one that relaxes this requirement provided you
```

Comment:
requiring ~~from~~ the implementation *type* to be `Send`

(adding *type* is of course not something grammatical, but given that "implementation" usually refers to a function, I think it makes sense for the type impl'ing `Implementation` [that trait has a really weird name...] to be referred to as "the implementation type")